File: dotnet_projects/OcrLiteOnnxVB/OcrLiteLib/AngleNet.vb
Imports System.Drawing
Imports Emgu.CV
Imports Emgu.CV.CvEnum
Imports Emgu.CV.Structure
Imports Microsoft.ML.OnnxRuntime
Imports Microsoft.ML.OnnxRuntime.Tensors

Namespace OcrLiteLib
    Public Class AngleNet
        Private ReadOnly MeanValues As Single() = {127.5F, 127.5F, 127.5F}
        Private ReadOnly NormValues As Single() = {1.0F / 127.5F, 1.0F / 127.5F, 1.0F / 127.5F}
        Private Const angleDstWidth As Integer = 192
        Private Const angleDstHeight As Integer = 32
        Private Const angleCols As Integer = 2
        Private angleNet As InferenceSession

        Public Sub New(path As String, numThread As Integer)
            Try
                Dim op As New SessionOptions()
                op.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_EXTENDED
                op.InterOpNumThreads = numThread
                op.IntraOpNumThreads = numThread
                angleNet = New InferenceSession(path)
            Catch ex As Exception
                Console.WriteLine(ex.Message + ex.StackTrace)
                Throw ex
            End Try
        End Sub

        Protected Overrides Sub Finalize()
            Try
                angleNet.Dispose()
            Finally
                MyBase.Finalize()
            End Try
        End Sub

        Public Function GetAngles(partImgs As List(Of Mat), doAngle As Boolean, mostAngle As Boolean) As List(Of Angle)
            Dim angles As New List(Of Angle)()
            If doAngle Then
                For i As Integer = 0 To partImgs.Count - 1
                    Dim startTicks = DateTime.Now.Ticks
                    Dim angle = GetAngle(partImgs(i))
                    Dim endTicks = DateTime.Now.Ticks
                    Dim angleTime = (endTicks - startTicks) / 10000.0F
                    angle.Time = angleTime
                    angles.Add(angle)
                Next
            Else
                For i As Integer = 0 To partImgs.Count - 1
                    Dim angle = New Angle(-1, 0, 0)
                    angles.Add(angle)
                Next
            End If
            'Most Possible AngleIndex
            If doAngle AndAlso mostAngle Then
                Dim angleIndexes As New List(Of Integer)
                angles.ForEach(Sub(x) angleIndexes.Add(x.Index))

                Dim sum As Double = angleIndexes.Sum()
                Dim halfPercent As Double = angles.Count / 2.0F
                Dim mostAngleIndex As Integer
                If sum < halfPercent Then
                    'all angle set to 0
                    mostAngleIndex = 0
                Else
                    'all angle set to 1
                    mostAngleIndex = 1
                End If
                Console.WriteLine("Set All Angle to mostAngleIndex({mostAngleIndex})")
                For i As Integer = 0 To angles.Count - 1
                    Dim angle As Angle = angles(i)
                    angle.Index = mostAngleIndex
                    angles(i) = angle
                Next
            End If
            Return angles
        End Function

        Private Function GetAngle(src As Mat) As Angle
            Dim angle As New Angle()
            Dim angleImg As Mat = AdjustTargetImg(src, angleDstWidth, angleDstHeight)
            Dim inputTensors As Tensor(Of Single) = OcrUtils.SubstractMeanNormalize(angleImg, MeanValues, NormValues)
            Dim inputs = New List(Of NamedOnnxValue)() From {
                NamedOnnxValue.CreateFromTensor("input", inputTensors)
            }
            Try
                Using results As IDisposableReadOnlyCollection(Of DisposableNamedOnnxValue) = angleNet.Run(inputs)
                    Dim resultsArray = results.ToArray()
                    Console.WriteLine(resultsArray)
                    Dim outputData As Single() = resultsArray(0).AsEnumerable(Of Single)().ToArray()
                    Return ScoreToAngle(outputData, angleCols)
                End Using
            Catch ex As Exception
                'throw ex;
                Console.WriteLine(ex.Message + ex.StackTrace)
            End Try
            Return angle

        End Function

        Private Function ScoreToAngle(srcData As Single(), angleCols As Integer) As Angle
            Dim angleIndex As Integer = 0
            Dim maxValue As Single = -1000.0F
            For i As Integer = 0 To angleCols - 1
                If i = 0 Then
                    maxValue = srcData(i)
                ElseIf srcData(i) > maxValue Then
                    angleIndex = i
                    maxValue = srcData(i)
                End If
            Next
            Dim angle As New Angle(angleIndex, maxValue)
            Return angle
        End Function

        Private Function AdjustTargetImg(src As Mat, dstWidth As Integer, dstHeight As Integer) As Mat
            Dim srcResize As New Mat()
            Dim scale As Single = CSng(dstHeight) / CSng(src.Rows)
            Dim angleWidth As Integer = CInt(CSng(src.Cols) * scale)
            CvInvoke.Resize(src, srcResize, New Size(angleWidth, dstHeight))
            Dim srcFit As New Mat(dstHeight, dstWidth, DepthType.Cv8U, 3)
            'srcFit.SetTo(new MCvScalar(255,255,255));
            If angleWidth < dstWidth Then
                CvInvoke.CopyMakeBorder(srcResize, srcFit, 0, 0, 0, dstWidth - angleWidth,
                    BorderType.Isolated, New MCvScalar(255, 255, 255))
            Else
                Dim rect As New Rectangle(0, 0, dstWidth, dstHeight)
                Dim partAngle As New Mat(srcResize, rect)
                partAngle.CopyTo(srcFit)
            End If
            Return srcFit
        End Function
    End Class
End Namespace

File: dotnet_projects/OcrLiteOnnxCs/OcrLiteLib/AngleNet.cs
using Emgu.CV;
using Emgu.CV.CvEnum;
using Emgu.CV.Structure;
using Microsoft.ML.OnnxRuntime;
using Microsoft.ML.OnnxRuntime.Tensors;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;

namespace OcrLiteLib
{
    class AngleNet
    {
        private readonly float[] MeanValues = { 127.5F, 127.5F, 127.5F };
        private readonly float[] NormValues = { 1.0F / 127.5F, 1.0F / 127.5F, 1.0F / 127.5F };
        private const int angleDstWidth = 192;
        private const int angleDstHeight = 32;
        private const int angleCols = 2;
        private InferenceSession angleNet;

        public AngleNet() { }

        ~AngleNet()
        {
            angleNet.Dispose();
        }

        public void InitModel(string path, int numThread)
        {
            try
            {
                SessionOptions op = new SessionOptions();
                op.GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_EXTENDED;
                op.InterOpNumThreads = numThread;
                op.IntraOpNumThreads = numThread;
                angleNet = new InferenceSession(path);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ex.StackTrace);
                throw ex;
            }
        }

        public List<Angle> GetAngles(List<Mat> partImgs, bool doAngle, bool mostAngle)
        {
            List<Angle> angles = new List<Angle>();
            if (doAngle)
            {
                for (int i = 0; i < partImgs.Count; i++)
                {
                    var startTicks = DateTime.Now.Ticks;
                    var angle = GetAngle(partImgs[i]);
                    var endTicks = DateTime.Now.Ticks;
                    var angleTime = (endTicks - startTicks) / 10000F;
                    angle.Time = angleTime;
                    angles.Add(angle);
                }
            }
            else
            {
                for (int i = 0; i < partImgs.Count; i++)
                {
                    var angle = new Angle();
                    angle.Index = -1;
                    angle.Score = 0F;
                    angles.Add(angle);
                }
            }
            //Most Possible AngleIndex
            if (doAngle && mostAngle)
            {
                List<int> angleIndexes = new List<int>();
                angles.ForEach(x => angleIndexes.Add(x.Index));

                double sum = angleIndexes.Sum();
                double halfPercent = angles.Count / 2.0f;
                int mostAngleIndex;
                if (sum < halfPercent)
                {//all angle set to 0
                    mostAngleIndex = 0;
                }
                else
                {//all angle set to 1
                    mostAngleIndex = 1;
                }
                Console.WriteLine($"Set All Angle to mostAngleIndex({mostAngleIndex})");
                for (int i = 0; i < angles.Count; ++i)
                {
                    Angle angle = angles[i];
                    angle.Index = mostAngleIndex;
                    angles[i] = angle;
                }
            }
            return angles;
        }

        private Angle GetAngle(Mat src)
        {
            Angle angle = new Angle();
            Mat angleImg = AdjustTargetImg(src, angleDstWidth, angleDstHeight);
            Tensor<float> inputTensors = OcrUtils.SubstractMeanNormalize(angleImg, MeanValues, NormValues);
            var inputs = new List<NamedOnnxValue>
            {
                NamedOnnxValue.CreateFromTensor("input", inputTensors)
            };
            try
            {
                using (IDisposableReadOnlyCollection<DisposableNamedOnnxValue> results = angleNet.Run(inputs))
                {
                    var resultsArray = results.ToArray();
                    Console.WriteLine(resultsArray);
                    float[] outputData = resultsArray[0].AsEnumerable<float>().ToArray();
                    return ScoreToAngle(outputData, angleCols);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message + ex.StackTrace);
                //throw ex;
            }
            return angle;

        }

        private Angle ScoreToAngle(float[] srcData, int angleCols)
        {
            Angle angle = new Angle();
            int angleIndex = 0;
            float maxValue = -1000.0F;
            for (int i = 0; i < angleCols; i++)
            {
                if (i == 0) maxValue = srcData[i];
                else if (srcData[i] > maxValue)
                {
                    angleIndex = i;
                    maxValue = srcData[i];
                }
            }
            angle.Index = angleIndex;
            angle.Score = maxValue;
            return angle;
        }

        private Mat AdjustTargetImg(Mat src, int dstWidth, int dstHeight)
        {
            Mat srcResize = new Mat();
            float scale = (float)dstHeight / (float)src.Rows;
            int angleWidth = (int)((float)src.Cols * scale);
            CvInvoke.Resize(src, srcResize, new Size(angleWidth, dstHeight));
            Mat srcFit = new Mat(dstHeight, dstWidth, DepthType.Cv8U, 3);
            //srcFit.SetTo(new MCvScalar(255,255,255));
            if (angleWidth < dstWidth)
            {
                CvInvoke.CopyMakeBorder(srcResize, srcFit, 0, 0, 0, dstWidth - angleWidth, BorderType.Isolated, new MCvScalar(255, 255, 255));
            }
            else
            {
                Rectangle rect = new Rectangle(0, 0, dstWidth, dstHeight);
                Mat partAngle = new Mat(srcResize, rect);
                partAngle.CopyTo(srcFit);
            }
            return srcFit;
        }

    }
}
